docs/build.md

```diff
@@ -0,0 +1,56 @@
+## Build and package configuration
+
+Build variables should be defined in a single file,
+so that anyone who wants to build Rancher needs to only edit this file to change configuration and dependency versions.
+
+Rancher relies on various subcomponents, such as the webhook.
+These typically need to have set versions for Rancher to build and run properly.
+Build variables can be used in different places and supplied to the applications in a variety of ways,
+including as environment variables in Dockerfiles, constants in Go code, and so on.
+
+The `config.yaml` is the single source of truth. It lists all values by name and value.
+Changes to it should be committed to source control.
+
+### Update an existing value
+
+Edit the config.yaml file and update the desired value.
+Then re-build Rancher with `make build` or re-package it with `make package`.
+
+### Add a new value
+
+The main idea is to pass a configuration value in one of two ways:
+
+1. As a `--build-arg` to `docker image build`, capture it as an `ARG` in the Dockerfile
+   via `docker build --build-arg MYVAR="$MYVAR" ...`
+2. As an injected variable to the Go binary at compile time via `go build -ldflags "-X main.MYVAR=$MYVAR ...`.
```

Comment:
Is there an example for go one ? I don't see in this PR. Overall this is interesting approach, and clever. I'm not sure it will scale well to other parts of the release process. For example we can't use env-vars to change things in setting.go, as env-vars are already treating as a second-class citizen there. I suppose the ldflags part will work, but that could get really messy if there are 50 ldflags, and taking the value out of setting.go will be confusing.  But for now thats ok, we can refactor once we get further along. 

> As as our Dockerfiles and Go files aren't structured data, we don't want to do any text manipulation with sed to replace text

The benefit to using `sed` or any other text manipulation tool is that you are guaranteed to not having problems at build or run time, everything is left in the developers hands and gets code-reviewed on the way in. I still want to consider that as an option to see how well this scales.

Reply:
No example of a Go variable yet, since we initially just wanted to port over the webhook version, which is an env var. But I can add one for something, like the shell version, which would be a variable.

Reply:
> The benefit to using sed or any other text manipulation tool is that you are guaranteed to not having problems at build or run time, everything is left in the developers hands and gets code-reviewed on the way in. I still want to consider that as an option to see how well this scales.

The only requirement in the current format is that your target value goes in the `config.yaml` - everything else is left up to how that specific value is persisted in Rancher (wherever that may be). This is all to say that sed is still an option for our build scripts.

That being said, I very much dislike sed based pipelines. I would disagree that they guarantee not having problems at build or run time - they can result in un-intended variables being changed (i.e. when replacing `CATTLE_RANCHER_WEBHOOK_VERSION` did you also update `CATTLE_RANCHER_WEBHOOK_MIN_VERSION` by accident?). I also think they are much worse in terms of read-ability.

But we can evaluate this as time goes by, we don't need to make a decision today.

Reply:
Agree, we can solve later. I'm going to leave this thread unresolved.